README.md
# Don't Starve Together mod: Type to Craft

## Problem
It is hard to find from the menu what you want to craft.

## Solution

Use typing for faster crafting.

Process: [Development Log](DEVLOG.md)

Release: [Steam Workshop](https://steamcommunity.com/sharedfiles/filedetails/?id=2711603423)

## References
 - `$DST_SCRIPTS/screens/chatinputscreen.lua`
 - mod: [Craft Hotkey](https://steamcommunity.com/sharedfiles/filedetails/?id=1920286184)
 - mod: [Slash Commands For Creative](https://steamcommunity.com/sharedfiles/filedetails/?id=1416903676)



CHANGELOG.md

# Version history
- 30 nu: fix scaling without popup
- 20 nu: recipe popup is configurable on/off
- 10 nu: add skin support to recipies
- 1 nu: show recipe popup
- 300 pu: fix predictions updated when entered uppercase
- 200 pu: fix plurals in requirements
- 100 pu: list all missed things
- 30 pu: marketing: new screenshot
- 20 pu: marketing: new description
- 10 pu: marketing: new icon
- 1 pu: item uses last used skin 
- 100 fu: remove chat log
- 50 fu: refactor: renaming and moving
- 40 fu: refactor: rm unused code
- 30 fu: refactor: generate itemlists on demand
- 20 fu: refactor: add helper module
- 10 fu: refactor: split code
- 2 fu: custom modifier
- 1 fu: configurable binding
- 100 au: ESC does not open menu when prompt open
- 10 au: f1 also closes prompt
- 9 au: doesn't open with other huds open
- 8 au: fix crash with empty strings
- 7 au: fix texture size to power of 2
- 6 au: last item remembered when crafted through prompt
- 5 au: show with icon if invent/build
- 4 au: reorganize tasks
- 3 au: fix placing bufferd items
- 2 au: fix crash on pressing enter
- 1 au: steam workshop release
- 300 zu: text can be entered also without #
- 200 zu: lowercase text
- 100 zu: choose predicted text directly
- 20 zu: lowercase predictive text
- 10 zu: search uses real item names
- 1 zu: prototype missing recipe
- 600 yu: fix missing recipes
- 500 yu: inform how many missing ingredients required
- 400 yu: debug command: reboot F5
- 300 yu: repeat last recipe F1
- 200 yu: add changelog
- 100 yu: inform missing ingredients
- 20 yu: inform if recipe not known
- 10 yu: predictive text
- 4 yu: inform when searched item not found
- 3 yu: add readme
- 2 yu: fix esc issue
- 1 yu: release on github
- 100 ru: show craft input with hotkey
- 10 ru: use command to create specific item
- 2 ru: craft spear on command
- 1 ru: copy paste mod as basis
- 200 qu: research basic modding info
- 100 qu: research reference mods
- 10 qu: change willow text to test modding
- 1 qu: unbundle scripts.zip


modinfo.lua
name = 'Type to Craft'
description = 'Finder for crafting. You can search and craft items with typing the recipe\'s name using only your keyboard. Allows building items quickly with predictive text and without having to navigate menus.\n\nCrafts the item with the skin that was last chosen.\n\n Default bindings:\n C = Open craft prompt.\n Shift+C = Repeat last craft.'
author = 'Sakea'
version = '0.000004'
forumthread = ''

api_version = 10
priority = 1

dont_starve_compatible = false
reign_of_giants_compatible = false
shipwrecked_compatible = false
dst_compatible = true

all_clients_require_mod = false
client_only_mod = true
--server_filter_tags = {}

icon_atlas = 'atlas.xml'
icon       = 'modicon.tex'


configuration_options =
{
  {
    name = "bind",
    label = "Open crafting input",
    hover = "Bind button to open crafting input",

    options = {
      { description = "C", data = "KEY_C" },
      { description = "F", data = "KEY_F" },
      { description = "G", data = "KEY_G" },
      { description = "R", data = "KEY_R" },
      { description = "X", data = "KEY_X" },
      { description = "F1 (F2 repeats)", data = "KEY_F1" },
    },

    default = "KEY_C"
  },
  {
    name = "modifier",
    label = "Repeat last crafting",
    hover = "Bind which modifier + Craft button repeats last crafting. ",

    options = {
      { description = "Shift", data = "KEY_SHIFT" },
      { description = "Ctrl", data = "KEY_CTRL"},
      { description = "Alt", data = "KEY_ALT" },
    },

    default = "KEY_SHIFT"
  },
  {
    name = "ui",
    label = "Crafting pop up",
    hover = "Show crafting pop up with item info.",

    options = {
      { description = "On", data = "full" },
--      { description = "Minimal", data = "minimal" },
      { description = "Off", data = "off" },
    },

    default = "full"
  },
}

modmain.lua
local Helper = require "helper"
local CraftInput = require "craftscreen"

Helper = Helper:new(GLOBAL)

local ICON_CRAFT = "󰀌" --hammer
local ICON_INVENT = "󰀏" --lightbulb
local ICON_CANT_CRAFT = ""

local KEY_DEBUG = GLOBAL.KEY_BACKSLASH
local KEY_RESET = GLOBAL.KEY_F5
local KEY_CRAFT_INPUT = GLOBAL.KEY_F1
local KEY_CRAFT_LAST = GLOBAL.KEY_F2
local KEY_ESCAPE = GLOBAL.KEY_ESCAPE
local KEY_ENTER = GLOBAL.KEY_ENTER

local bind = GetModConfigData("bind")
local binded_modifier = GetModConfigData("modifier")
local ui_type = GetModConfigData("ui")

KEY_CRAFT_INPUT = GLOBAL[bind]

state = {lastItem = "",skin=nil}
 
local function isPlayerAvailable()
	local DST = GLOBAL.TheSim:GetGameID() == "DST"
	return DST and GLOBAL.ThePlayer ~= nil
end

local function getPlural(msg, amount)
	if amount > 1 and string.sub(msg, -1) == "s" then
		return amount.." "..msg
	elseif amount > 1 then
		return amount.." "..msg.."s"
	else
		return msg
	end
end

local function craftItem(recipeName, skin)
	
	if not isPlayerAvailable() then
		return
	end
	
	local recipe = GLOBAL.GetValidRecipe(recipeName)
	
	local talker = GLOBAL.ThePlayer.components.talker
	local builder = GLOBAL.ThePlayer.replica.builder

	if recipe == nil then
		Helper:sendMessage(ICON_CANT_CRAFT.." No such item.")
		return
	end

	if skin ~= nil then
		GLOBAL.Profile:SetLastUsedSkinForItem(recipeName, skin)
	end

	state.lastItem = recipeName

	local icon = ICON_CRAFT
	local localizedRecipeName = Helper:getReadableItemName(recipe.name)
	local invented_msg = ""
	local missing_msg = ""

	if not builder:KnowsRecipe(recipeName) then
		if  GLOBAL.CanPrototypeRecipe(recipe.level, builder:GetTechTrees()) and
			builder:CanLearn(recipe.name) then
			icon = ICON_INVENT
		else
			invented_msg = "I don't know the recipe for "..localizedRecipeName.."."
		end
	end

	if not builder:IsBuildBuffered(recipeName) and not builder:CanBuild(recipeName) then
		for i, v in ipairs(recipe.ingredients) do
			if not builder.inst.replica.inventory:Has(v.type, math.max(1, GLOBAL.RoundBiasedUp(v.amount * builder:IngredientMod()))) then
				missing_msg = missing_msg .." "..getPlural(v.type, v.amount) ..","
			end
		end
		for i, v in ipairs(recipe.character_ingredients) do
			if not builder:HasCharacterIngredient(v) then
				missing_msg = missing_msg .." ".. v.type..","
            end
        end
        for i, v in ipairs(recipe.tech_ingredients) do
            if not builder:HasTechIngredient(v) then
				missing_msg = missing_msg .." tech ingredient."
            end
        end
		--Helper:sendMessage(ICON_CANT_CRAFT.." "..localizedRecipeName.."? Something wrong.")
	end

	if invented_msg ~= "" and missing_msg == ""  then
		Helper:sendMessage(ICON_CANT_CRAFT.." "..invented_msg)
	elseif invented_msg == "" and missing_msg ~= ""  then
		Helper:sendMessage(ICON_CANT_CRAFT.." "..localizedRecipeName.."? I don't have"..string.sub(missing_msg, 1, -2))
	elseif invented_msg ~= "" and missing_msg ~= ""  then
		Helper:sendMessage(ICON_CANT_CRAFT.." "..invented_msg .." Also I don't have"..string.sub(missing_msg, 1, -2))
	else
		Helper:sendMessage(icon.." Crafting "..localizedRecipeName)
		
		if skin == nil then
			skin = GLOBAL.Profile:GetLastUsedSkinForItem(recipeName)
		end
		
		if recipe.placer == nil then
			return builder:MakeRecipeFromMenu(recipe, skin)
		else
			if not builder:IsBuildBuffered(recipeName) then
				builder:BufferBuild(recipeName)
			end
			return GLOBAL.ThePlayer.components.playercontroller:StartBuildPlacementMode(recipe, nil)
		end
	end
end

local function closePrompt()
	local activeScreen = TheFrontEnd:GetActiveScreen()

	if activeScreen.name == "CraftInput" then
		activeScreen:Close()
		return true
	end
end

local function craftLast()
	
	if state.lastItem ~= "" then
		craftItem(state.lastItem, state.skin)
	end
end

local function startInput()
	if not isPlayerAvailable() then
		return
	end

	if GLOBAL.TheInput:IsKeyDown(GLOBAL[binded_modifier]) then
		craftLast()
		return
	end

	local activeScreen = TheFrontEnd:GetActiveScreen()

	if activeScreen.name == "CraftInput" then
		--pass through - we are in middle of input
		return false
	elseif activeScreen.name == "HUD" then
		-- no menus open
		TheFrontEnd:PushScreen(CraftInput(GLOBAL, craftItem, ui_type))
	end
end

local function enterKey()
	
	local activeScreen = TheFrontEnd:GetActiveScreen()
	
	if activeScreen.name ~= "CraftInput" then
		return
	end
	
	if Helper:cleanItemName(activeScreen.chat_edit:GetString()) == "" then
		activeScreen:Close()
		return
	end
	
	item = activeScreen:GetSelectedItem()
	if item then
		activeScreen:Close()
		craftItem(item)
	end
end

GLOBAL.TheInput:AddKeyDownHandler(GLOBAL[bind], function() startInput() end)
GLOBAL.TheInput:AddKeyUpHandler(KEY_ESCAPE, function() return closePrompt() end)
GLOBAL.TheInput:AddKeyUpHandler(KEY_ENTER, function() enterKey() end)

if bind == "KEY_F1" then
	GLOBAL.TheInput:AddKeyDownHandler(KEY_CRAFT_LAST, function() craftLast() end)
end

scripts/common.lua
itemsByID={}
itemsByName = {}

function sendMessage(msg)
	--function Talker:Say(script, time, noanim, force, nobroadcast, colour, text_filter_context, original_author_netid)
	GLOBAL.ThePlayer.components.talker:Say(msg, nil, nil, nil, nil)
end

function getItemName(recipeName)
	return GLOBAL.STRINGS.NAMES[string.upper(recipeName)]
end
 
local function getItemNames()
	local builder = GLOBAL.ThePlayer.replica.builder

	local n=0
	
	for k,v in pairs(GLOBAL.AllRecipes) do
		local recipe = GLOBAL.GetValidRecipe(k)
		name = getItemName(k)
		if name ~= nil and recipe ~= nil then
			--if builder:KnowsRecipe(k) or GLOBAL.CanPrototypeRecipe(recipe.level, builder:GetTechTrees()) then
				n=n+1
				itemsByID[n]=string.lower(name)
				itemsByName[string.lower(name)] = k
			--end
		end
	end

	return itemsByID
end


getItemNames()

scripts/helper.lua

local Helper = {}

function Helper:new(global)
    self.readableItemNames={}
    self.rawItemNames = {}
    self._G = global

    return self

end

function Helper:cleanItemName(name)
	name = string.lower(name)
	name = name:match( "^%s*(.-)%s*$" )
	return name:match( "^#*(.-)#*$" )
end

function Helper:sendMessage(msg)
	--function Talker:Say(script, time, noanim, force, nobroadcast, colour, text_filter_context, original_author_netid)
	self._G.ThePlayer.components.talker:Say(msg, 1+(#msg)/20, nil, nil, nil)
end

function Helper:getRawItemName(readableName)

    if #(self.rawItemNames) == 0 then
        self.generateItemNames()
    end

    return self.rawItemNames[readableName]
end

function Helper:getReadableItemName(recipeName)
	return self._G.STRINGS.NAMES[string.upper(recipeName)]
end

function Helper:listReadableItemNames()
    if #(self.rawItemNames) == 0 then
        self:generateItemNames()
    end
    
    return self.readableItemNames
end

function Helper:generateItemNames()
    if self == nil then
        return
    end

	local builder = self._G.ThePlayer.replica.builder

	local n=0
	
	for k,v in pairs(self._G.AllRecipes) do
		local recipe = self._G.GetValidRecipe(k)
		local readableName = self:getReadableItemName(k)
		if readableName ~= nil and recipe ~= nil then
			--if builder:KnowsRecipe(k) or self._G.CanPrototypeRecipe(recipe.level, builder:GetTechTrees()) then
				n=n+1
				self.readableItemNames[n]=string.lower(readableName)
				self.rawItemNames[string.lower(readableName)] = k
			--end
		end
	end

	return self.readableItemNames
end


return Helper

scripts/craftscreen.lua
require "util"

local TextCompleter = require "util/textcompleter"
local Screen = require "widgets/screen"
local RecipePopup = require "widgets/recipepopup"
local RecipeTile = require "widgets/recipetile"
local TextEdit = require "widgets/textedit"
local Widget = require "widgets/widget"
local Text = require "widgets/text"
local ImageButton = require "widgets/imagebutton"
--local VirtualKeyboard = require "screens/virtualkeyboard"
local Emoji = require("util/emoji")
local UserCommands = require("usercommands")

local Helper = require "helper"
local GLOBAL

local CraftInput = Class(Screen, function(self, glob, craft_fn, ui_type)
	Screen._ctor(self, "CraftInput")
	self._G = glob
    GLOBAL = glob
    self.craftItem = craft_fn
	self.runtask = nil
    self.ui_type = ui_type
	self.is_crafting_input = true
	Helper = Helper:new(self._G)

	self:DoInit()
end)

function CraftInput:OnBecomeActive()
    CraftInput._base.OnBecomeActive(self)

    self.chat_edit:SetFocus()
    self.chat_edit:SetEditing(true)

	if self._G.IsConsole() then
		TheFrontEnd:LockFocus(true)
	end

end

function CraftInput:OnBecomeInactive()
    CraftInput._base.OnBecomeInactive(self)

    if self.runtask ~= nil then
        self.runtask:Cancel()
        self.runtask = nil
    end
end

function CraftInput:Run()
    local chat_string = self.chat_edit:GetString()
	local item = Helper:cleanItemName(chat_string)
	
	self.craftItem(Helper:getRawItemName(item))
    chat_string = chat_string ~= nil and chat_string:match("^%s*(.-%S)%s*$") or ""
    if chat_string == "" then
        return
    end
end

function CraftInput:Close()
    self._G.TheInput:EnableDebugToggle(true)
    TheFrontEnd:PopScreen(self)
    return true
end

local function DoRun(inst, self)
    self.runtask = nil
    self:Run()
    self:Close()
end

local function DoClose(inst, self)
    self.runtask = nil
    self:Close()
end

function CraftInput:OnTextEntered()
    if self.runtask ~= nil then
        self.runtask:Cancel()
    end
    self.runtask = self.inst:DoTaskInTime(0, DoRun, self)
end

function CraftInput:DoInit()
    self._G.TheInput:EnableDebugToggle(false)

    local label_height = 50
    local fontsize = 30
    local edit_width = 850
    local edit_width_padding = 0
    local chat_type_width = 150

    self.root = self:AddChild(Widget("chat_input_root"))

    self.root:SetScaleMode(self._G.SCALEMODE_PROPORTIONAL)
    self.root:SetHAnchor(self._G.ANCHOR_MIDDLE)
    self.root:SetVAnchor(self._G.ANCHOR_BOTTOM)
    self.root = self.root:AddChild(Widget(""))

    self.root:SetPosition(45.2, 100, 0)

	if not self._G.TheInput:PlatformUsesVirtualKeyboard() then
	    self.chat_type = self.root:AddChild(Text(self._G.TALKINGFONT, fontsize))
	    self.chat_type:SetPosition(-505, 0, 0)
	    self.chat_type:SetRegionSize(chat_type_width, label_height)
	    self.chat_type:SetHAlign(self._G.ANCHOR_RIGHT)
	    self.chat_type:SetString("Craft item:")
	    self.chat_type:SetColour(.6, .6, .6, 1)
	end

    self.chat_edit = self.root:AddChild(TextEdit(self._G.TALKINGFONT, fontsize, ""))
    self.chat_edit.edit_text_color = self._G.WHITE
    self.chat_edit.idle_text_color = self._G.WHITE
    self.chat_edit:SetEditCursorColour(self._G.unpack(self._G.WHITE))
    self.chat_edit:SetPosition(-.5 * edit_width_padding, 0, 0)
    self.chat_edit:SetRegionSize(edit_width - edit_width_padding, label_height)
    self.chat_edit:SetHAlign(self._G.ANCHOR_LEFT)

    -- the screen will handle the help text
    self.chat_edit:SetHelpTextApply("")
    self.chat_edit:SetHelpTextCancel("")
    self.chat_edit:SetHelpTextEdit("")
    self.chat_edit.HasExclusiveHelpText = function() return false end

    self.chat_edit.OnTextEntered = function() self:OnTextEntered() end
    self.chat_edit:SetPassControlToScreen(self._G.CONTROL_CANCEL, true)
    self.chat_edit:SetPassControlToScreen(self._G.CONTROL_MENU_MISC_2, true) -- toggle between say and whisper
    self.chat_edit:SetPassControlToScreen(self._G.CONTROL_SCROLLBACK, true)
    self.chat_edit:SetPassControlToScreen(self._G.CONTROL_SCROLLFWD, true)
    self.chat_edit:SetTextLengthLimit(self._G.MAX_CHAT_INPUT_LENGTH)
    self.chat_edit:EnableWordWrap(false)
    self.chat_edit:EnableRegionSizeLimit(true)
    self.chat_edit:EnableScrollEditWindow(false)
    
    local ci = self
	self.chat_edit:SetForceEdit(true)

    local CloseMaybe = function()
        if ci.recipe.shown == false then
            ci:Close()
        end
        ci.chat_edit:Hide() ci.chat_type:Hide() self.chat_edit:SetForceEdit(false) ci.chat_edit:SetEditing(false)
    end
    self.chat_edit.OnStopForceEdit = CloseMaybe

    -- Word Prediction --

    local prediction_dictionary = {
        words = Helper:listReadableItemNames(),
		delim = "#",
		--GetDisplayString = function(word) return word end
	}
    
    self.chat_edit:EnableWordPrediction({width = 800, mode=self._G.Profile:GetChatAutocompleteMode()})
	self.chat_edit:AddWordPredictionDictionary(prediction_dictionary)
    self.chat_edit:SetString("#")

	self.chat_edit.OnTextInputted = function(text, k)
		self.chat_edit:SetString(string.lower(self.chat_edit:GetString()))
        self.chat_edit.prediction_widget:RefreshPredictions()
		return false
	end

    self.chat_edit.prediction_widget.GetSelectedItem = function(self)
        if #(self.prediction_btns) > 0 then
            local strItem = self.prediction_btns[self.active_prediction_btn].text.string
            
            -- Remove #
            strItem = strItem:match( "^#*(.-)#*$" )
            return Helper:getRawItemName(strItem)
        end
    end

    -- Crafting Pop-up

    --self:AddChild(Image("images/hud.xml", "craft_bg.tex"))
    self.slot = self.root:AddChild(Image("images/hud.xml", "craft_slot.tex"))
    self.tile = self.root:AddChild(RecipeTile(nil))
    self.recipe = self.root:AddChild(RecipePopup(false))
    self.slot:SetPosition(-565, 270)
    self.tile:SetPosition(-560, 273, 0)
    self.recipe:SetPosition(-540, 250, 0)
    self.slot:SetScale(0.75, 0.75)
    self.tile:SetScale(0.75, 0.75)
    self.recipe:SetScale(0.75, 0.75)
    self.slot:Hide()
    self.recipe:Hide()
    self.recipe.OnLoseFocus = function()
        ci.recipe:Hide() ci.slot:Hide() ci.tile:Hide()
        ci.chat_edit:Show() ci.chat_type:Show()
        ci.chat_edit:SetForceEdit(true) ci.chat_edit:SetEditing(true) ci.chat_edit:SetFocus()
    end
    
    --self.craft_name = self.root:AddChild(Text(self._G.UIFONT, fontsize))
    --self.craft_name:SetPosition(-300, 80)
    --self.craft_name:SetHAlign(self._G.ANCHOR_LEFT)
	
    local full_ui = function(self, key, down)
        local res = self:origOnRawKey(key, down)
        if self["prediction_btns"] then

            local buildIt = function() 

                local skin = nil
                if ci.recipe["skins_spinner"] then
                    skin = ci.recipe.skins_spinner.GetItem()
                end
                ci.craftItem(ci.recipe.recipe.name, skin)
                ci.inst:DoTaskInTime(0, DoClose, ci)
                return true
            end

            ci.recipe.button:SetWhileDown(buildIt)
            ci.recipe.button:SetOnClick(buildIt)

            local item = self:GetSelectedItem()
            local recipe = GLOBAL.GetValidRecipe(item)
            if recipe == nil then
                ci.recipe:Hide()
                ci.tile:Hide()
                ci.slot:Hide()
                --ci.craft_name:SetString("")
            else
                ci.recipe:SetRecipe(recipe, ci._G.ThePlayer)
                --ci.recipe.button:Hide()
                ci.recipe:Show()
                ci.slot:Show()
                ci.tile:SetRecipe(recipe)
                ci.tile:Show()
                --ci.craft_name:SetString(Helper:getReadableItemName(item))
            end
        end

        return res
    end

    if self.ui_type == "full" then
        self.chat_edit.prediction_widget.origOnRawKey = self.chat_edit.prediction_widget.OnRawKey
        self.chat_edit.prediction_widget.OnRawKey = full_ui
    end
end

function CraftInput:GetSelectedItem()
	local pred_widget = self.chat_edit.prediction_widget
	return pred_widget:GetSelectedItem()
end


return CraftInput